bin/format-array.rb

```diff
@@ -0,0 +1,178 @@
+require 'json'
+
+# format-array expresses (and enacts) our per-file preferences for array formatting,
+# in cases where the preference differs from prettier.
+#
+# It has only one mode of operation:
+# Run it without any arguments to format the files.
+
+# format-array was written because we want to keep files consistently formatted,
+# while also optimising for human readability.
+#
+# As part of optimising for human readability,
+# some arrays should be formatted such that each element is on its own line,
+# whereas some other arrays should be formatted such that the entire array is on one line.
+#
+# We could not find an existing tool that allows us to specify array formatting how we wanted.
+
+# The configuration of array formatting preferences.
+#
+# format can be the following choices:
+#
+# * single_line (the entire array and all of its elements should be on a single line)
+# * multi_line (each element of the array is on its own line)
+# * multi_line_unless_single (like multi_line, except arrays with one element remain on a single line)
+# * multi_line_deep (multi_line, even applied to arrays within that array)
+# * padded (14 elements per line padded to 3 characters per element)
+#     (TODO: padding amount could be configurable, but we haven't needed it)
+formats = {
+  'book-store' => {
+    'basket' => :single_line,
+  },
+  'bowling' => {
+    'previousRolls' => :single_line,
+  },
+  'change' => {
+    'expected' => :single_line,
+  },
+  'connect' => {
+    'board' => :multi_line,
+  },
+  'diamond' => {
+    'expected' => :multi_line,
+  },
+  'dominoes' => {
+    'dominoes' => :single_line,
+  },
+  'flatten-array' => {
+    'array' => :multi_line_deep,
+    'expected' => :multi_line,
+  },
+  'forth' => {
+    'instructions' => :multi_line_unless_single,
+  },
+  'go-counting' => {
+    'board' => :multi_line,
+  },
+  'minesweeper' => {
+    'minefield' => :multi_line_unless_single,
+    'expected' => :multi_line_unless_single,
+  },
+  'ocr-numbers' => {
+    'rows' => :multi_line,
+  },
+  'rectangles' => {
+    'strings' => :multi_line_unless_single,
+  },
+  'saddle-points' => {
+    'matrix' => :multi_line,
+  },
+  'scale-generator' => {
+    'expected' => :single_line,
+  },
+  'sieve' => {
+    'expected' => :padded,
+  },
+  'transpose' => {
+    'lines' => :multi_line,
+    'expected' => :multi_line,
+  },
+  'variable-length-quantity' => {
+    'integers' => :single_line,
+    'expected' => :single_line,
+  },
+  'word-search' => {
+    'grid' => :multi_line,
+    'wordsToSearchFor' => :multi_line,
+  },
+}.each_value(&:freeze).freeze
+
+def single_line_arrays(contents, key)
+  # matches things of the form "key": [1, 2, 3]
+  # because this is not a multi-line regex,
+  # . does NOT match newlines.
+  contents.scan(/^( +)"#{key}": (\[.*\],?$)/)
```

Comment:
Theoretically speaking, this (using a regex) could also match the text in a comment, but that is both highly unlikely and pretty noticeable in a PR.

Reply:
This is a fascinating case to consider. Presumably it would be something like

```json
{
  "comment": [
    "consider the following input",
    "foo\": [1, 2, 3]"
  ]
}
```

That doesn't particularly look good in JSON, and for this to match you have to pass a `key` that includes a `\`... it's indeed pretty noticeable on both ends.